neonvm: build and pass kernel tools to neonvm runner

iddm opened the pull request:
This allows the compute instances to have the `tools` (for now just `perf` and the `kernel headers`) available in the guest VM, so that the VM can use those. These `tools` are required to be passed through only because they are tied to the kernel version they were built for. However, for the future, we are not limited to only such a use case and can basically include anything.

mikhail-sakhnov commented on neonvm-kernel/Dockerfile:
```diff
@@ -55,27 +74,55 @@ RUN set -e \
+    && echo "Building tools" \
+    && cd tools/perf \
+    && make LDFLAGS="-static -L/usr/lib/aarch64-linux-gnu -Wl,--whole-archive -lzstd -Wl,--no-whole-archive" NO_LIBPERL=1 NO_GTK2=1 NO_SDT=1 NO_JEVENTS=1
+    #&& make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- LDFLAGS=-static NO_LIBPERL=1 NO_GTK2=1 NO_SDT=1 NO_JEVENTS=1
```
leftover?

petuhovskiy commented on neonvm-kernel/Dockerfile:
```diff
@@ -25,7 +25,26 @@ RUN apt-get update && apt-get -y install \
     cpio \
     zstd \
-    libncurses-dev
+    libncurses-dev \
+    clang \
+    llvm-dev \
+    libdw-dev \
+    libslang2-dev \
+    libbfd-dev \
+    libiberty-dev \
+    libcap-dev \
+    libunwind-dev \
+    zlib1g-dev \
+    libzstd-dev \
+    liblzma-dev \
+    libaio-dev \
+    libtraceevent-dev \
+    libpfm4-dev \
+    systemtap-sdt-dev \
+    musl \
+    musl-dev \
+    musl-tools \
+    pkg-config
```
Are all these packages required for building perf?

petuhovskiy reviewed (commented):
LGTM

Would be nice to try running this `perf` on staging, both on x86/ARM, for postgres 15/16/17, to confirm that it's working in all possible configurations.

mikhail-sakhnov commented on neonvm-kernel/Dockerfile:
```diff
@@ -25,7 +25,26 @@ RUN apt-get update && apt-get -y install \
     cpio \
     zstd \
-    libncurses-dev
+    libncurses-dev \
```
should we introduce a separate build phase to make cleaner dependency list?

iddm replied:
Done. Please re-review.

iddm commented on neonvm-kernel/Dockerfile:
```diff
@@ -54,37 +54,107 @@ RUN set -e \
+    musl-tools \
+    pkg-config
+
+FROM build-kernel-deps AS build-tools-deps-arm64
+RUN dpkg --add-architecture arm64 && \
```
There are issues because of cross-compiling, so we need to install the arm64 headers and libraries for arm64 `perf` to build width and link against.

iddm replied:
I understand your frustration. I was there myself. I don't know. If you want me to recheck, sure.

iddm replied:
I think because we run docker in a VM and it doesn't care about the architecture on the host, it is still x86, so this is why we need to add the arm64 architecture on our own, and this is why we actually cross-compile and not just build the binaries for both, linux kernel and perf.

sharnoff replied:
I've now tested locally trying to cross-compile ARM -> x86, and it fails with
```
Makefile.config:458: *** No static glibc found, please install glibc-static.  Stop.
```

sharnoff replied:
It was the latest commit on this PR, testing with `make TARGET_ARCH=amd64 kernel` from an ARM system.

iddm commented on neonvm-runner/cmd/disks.go:
```diff
@@ -175,13 +185,18 @@ func createISO9660runtime(
+	// Add tools.
+	mounts = append(mounts, "/neonvm/bin/mkdir -p /mnt/tools")
+	mounts = append(mounts, "/neonvm/bin/mount -t iso9660 -o ro,exec $(/neonvm/bin/blkid -L vm-tools) /mnt/tools")
+	mounts = append(mounts, "/neonvm/bin/ln -s /mnt/tools/* /usr/local/bin/")
```
It is already there and works, it has been tested.

The best way for the compute to work is **without** having any knowledge about where perf is, or have a way to be configured to always know where it is precisely. I talked with Arthur about that and he suggested we just put it into the already-set `$PATH`.

iddm commented on neonvm-kernel/.gitignore:
```diff
@@ -3,3 +3,4 @@ vmlinuz
 # Exclude files for local kernel development.
 # Otherwise certain git operations become unexpectedly very expensive.
 linux/
+tools/
```
Done.

iddm commented:
Added kernel headers.

sharnoff reviewed (approved):
lgtm!

iddm commented:
In the known place will require this as well:

```diff
-    ignore=["autoscaling/neonvm-kernel/vmlinuz"],
+    ignore=["autoscaling/neonvm-kernel/vmlinuz", "autoscaling/neonvm-kernel/tools"],
```

iddm commented:
As we have changed in the compute to also allow using eBPFs, as well as there has been a long-awaiting request to support eBPFs fully, there are a bit more changes required now. In my last commit which I have just pushed, I have enabled the required kernel configuration options to enable us to use eBPFs at runtime successfully. The eBPFs now compile just fine, as some of those even work, but not all of them. This is due to the reason that some of the bpf headers require a kernel header file `GEN-reg-for-each.h` or something like that. This header is only generated for the x86 architecture and is not required for aarch64, so the latter should just work.

To pack the "tools", which now includes:

1. `perf` binary.
2. kernel headers and modules necessary for eBPFs and other kernel tools.

I had to make an adjustment in the ISO9660-generating code and change the version suffix of the kernel release name. First, the code for ISO9660 generation, which had existed before my changes, didn't recursively go through the `contentPath` and didn't package anything what isn't a file. In order to simplify the things, I made it work recursively over all the directories in the `contentPath` to include every single file from there, so to fully pack the `contentPath` into an `iso`-file. However, due to our naming of the kernel release, it wasn't possible to pack the things into the `iso` file as-is, as the ISO standard require no more than 31 characters per path entry, and our kernel release name for twice or even more than that. So I changed the version suffix to a much smaller one, still helpful, and also got approvals from the others from the autoscaling team.

I am going to leave for a couple of weeks now, so this is a list of the things i'll do once I come back:

1. Change the arm64 kernel configuration file accordingly, to enable the same features as I have enabled for x86.
4. Make sure the x86 required headers are generated and properly copied into the modules tree.

The testing can be done by `tilt up full`, and using the `autoscaling` changes and the `compute` changes.
